R/module_teal.R

```diff
@@ -136,6 +138,12 @@ srv_teal <- function(id, modules, teal_data_rv, filter = teal_slices()) {
       title = "SessionInfo"
     )
 
+    teal.widgets::verbatim_popup_srv(
+      "lockFile",
+      verbatim_content = create_lockfile(),
```

Comment:
@gogonzo it's few (2-5) seconds for this app, but we can do microbenchmarks benchmarks

Reply:
I checked for teal.osprey on my relatively quick machine it takes 8,5 sek. This is too much, I suggest to move it to the `init` and not to block other processes (maybe `future`?)

Reply:
Problem with `parallel` computations is that they have different architecture for every operating system. I do not see `parallel::mcparallel` function on Windows. Parallel computation involves launching worker processes, and you might not always have an access to do so. This snippet of the code would need to look different depending on the operating system.

Reply:
True, `parallel::mcparallel` is not available on windows. Can you explore other ways?

Reply:
maybe `callr::r_bg` then?

Reply:
Yup, let's try @gogonzo :) this is really good hint, to get this long processing stuff to run simultenously during the initizlizaion

Reply:
Just a note that we need to pass packages to a separate process so it represents the current shiny session. So we need to figure out what runs in shiny `teal::init`, then pass those to a separate process and then run `pak::lockfile_create`

Reply:
How about a newly released `ExtendedTask`? This seems to be a perfect use case of this. It might require vbump shiny dep but I think that's okey

Reply:
@m7pr this don't have to be done per shiny session - it (file) can be recreated once in init. We don't need to re-create file in shiny session, as package dependencies remain the same between sessions. `lockfile` doesn't represent current shiny session but dependencies which are needed to run the app

Reply:
I was able to incorporate `callr::r_bg()` solution https://github.com/insightsengineering/teal/tree/479_lockfile_callr%40main but it fails for `pak::lockfile_create` as it is unable to recreate the session that runs for shiny. Maybe I could improve it with `r_copycat`

> r_bg	Evaluate an expression in another R session, in the background
> r_copycat	Run an R process that mimics the current R process

Reply:
@pawelru I haven't heard about `ExtendedTask` - will try it out

Reply:
> @m7pr this doesn't have to be done per shiny session - it (file) can be recreated once in init. We don't need to re-create file in shiny session, as package dependencies remain the same between sessions. lockfile doesn't represent current shiny session but dependencies which are needed to run the app

**Below needs to be confirmed:**
Yeah, but if you run the `pak::lockfile_create` you get dependencies of what? Dependencies for a specific package, dependencies for a specific directory, or a specific session?  It will call the `pak::lockfile_create` for the directory but it has different behavior if a directory is a package. So for the app started inside a directory containing DESCRIPTION file it will just provide dependencies for this main package, however if DESCRIPTION is not in there it will run it for the whole directory - this is what we want. The whole directory or the whole session.

Reply:
I actually though `pak::lockfile_create` works the way `renv::snapshot()` works, which is the session snapshot. However `pak::lockfile_create` by default seeks for a package. You can specify packages on your own, with a GitHub reference or a custom remote repository reference, but you need to extract packages information on your own

<img width="520" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/59388ee0-144d-4085-b93e-eeb7da9e169c">

Reply:
Alrighty, so I think `callr::r_bg()` will work, if we detect package reference with `devtools::session_info()` and properly paste those to `pak::lockfile_create`. But that would work for the current session. If we do not have the session information because `callr::r_bg()` runs a new one, we need to have a way to figure out packages in a directory, and then their reference (source).

Reply:
**ExtendTask** https://rstudio.github.io/shiny/reference/ExtendedTask.html

> In normal Shiny reactive code, whenever an observer, calc, or output is busy computing, it blocks the current session from receiving any inputs or attempting to proceed with any other computation related to that session.

> The ExtendedTask class allows you to have an expensive operation that is started by a reactive effect, and whose (eventual) results can be accessed by a regular observer, calc, or output; but during the course of the operation, the current session is completely unblocked, allowing the user to continue using the rest of the app while the operation proceeds in the background.

Sounds like we could use it to run this `ExtendedTask` just at the beginning of `teal` app session. If the Task does not finish before someone clicks the button, then you can force the user/app to wait `until` the Task is finished.

Reply:
> Sounds like we could use it to run this `ExtendedTask` just at the beginning of `teal` app session. If the Task does not finish before someone clicks the button, then you can force the user/app to wait `until` the Task is finished.

What if one is not interested in the lock file at all? You will unnecessarily execute some lengthy operation. I was thinking more about sending a notification to the user interface upon click and unhide / insert download button when finished.
That's just my initial idea - thought I'm going to share it with you. Please experiment and share your thoughts. A similar problem is present for teal.reporter when downloading the report. We might re-use a similar approach in there.

Reply:
Yeah, so this could be the last question: whether calculation starts during the app initialization or on an on-click event.
I implemented `future`+`shiny::ExtendedTask` in here and I am currently testing it https://github.com/insightsengineering/teal/tree/479_lockfile_extendtask%40main

I also extended the `pak` approach, in which we need to specify packages (and their sources) to be included in a `.lock` file. Previously I though it figures out the session/dependencies on it's own, but it is only capable to figure out dependencies on it's own for a package-tree structure.

Reply:
Based on https://rstudio.github.io/shiny/reference/ExtendedTask.html `shiny::ExtendedTask()` needs to be `invoked` to get the process starting - I can only imagine this is invoked on an on-lick event. If this is invoked in `teal` `server` part alone, then the UI does not show until the computations are finished.

There is also a chapter on `ExtendedTask versus asynchronous reactives` and whyt it is better than using `promises` alone if you are curious. https://rstudio.github.io/shiny/reference/ExtendedTask.html

Reply:
Hey, I created an alternative PR https://github.com/insightsengineering/teal/pull/1232 where `pak::lockfile_create()` is run in a parallel process to the shiny session using `futures` + `shiny::ExtendedTask()`